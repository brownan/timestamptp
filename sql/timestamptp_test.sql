CREATE EXTENSION timestamptp;

SET DateStyle TO Postgres;
SET TimeZone TO EST;

--
-- See that the timezone offsets are preserved to minute resolution
SELECT timestamptp_in('2021-01-02T03:04:05-0123');

--
-- Use cast syntax
SELECT '2010-01-02 03:04:05 +06:07'::timestamptp;
SELECT '2000-01-01 00:00:00Z'::timestamptp;

--
-- Extract timestamptz and offset from a timestamptp
SELECT get_timestamp('2000-01-01 00:00:00 -07:00'::timestamptp);
SELECT get_offset('2000-01-01 00:00:00 -07:00'::timestamptp);

--
-- Construct a timestamptp from a timestamptz and an offset
SELECT make_timestamptp('2000-01-01 00:00:00 +06:00'::timestamptz, -300);

--
-- Error cases with timezone offset overflow
SELECT '2000-01-01 00:00:00 +16'::timestamptp;
SELECT '2000-01-01 00:00:00 -16'::timestamptp;
SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, -960);
SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, 960);

--
-- Just under the overflow
SELECT '2000-01-01 00:00:00 +15:59'::timestamptp;
SELECT '2000-01-01 00:00:00 -15:59'::timestamptp;
SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, -959);
SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, 959);

--
-- Cast a timestamptp to timestamptz
SELECT ('2000-01-01 00:00:00 +07:00'::timestamptp)::timestamptz;

--
-- Cast a timestamptz to timestamptp using implicit local timezone
SELECT ('2000-01-01 00:00:00 +00:00'::timestamptz)::timestamptp;

--
-- Tests for converting timestamptz to timestamptp in a DST-aware timezone
-- These two timestamps are on different sides of a DST transition, so the result
-- should have different offsets
SELECT make_timestamptp('2022-03-12 12:00:00Z'::timestamptz, 'America/New_York');
SELECT make_timestamptp('2022-03-13 12:00:00Z'::timestamptz, 'America/New_York');

--
-- Tests make_timestamptp() using other formats for specifying a timezone
-- as text. Offset spec, timezone abbreviations, and full timezone names should
-- be accepted.
SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, '-05:00'::text);
SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, '+0500'::text);
SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, 'EST'::text);
SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, 'EDT'::text);
SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, 'PST'::text);
SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, 'US/Eastern'::text);
