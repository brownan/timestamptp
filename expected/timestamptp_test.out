CREATE EXTENSION timestamptp;
SET DateStyle TO Postgres;
SET TimeZone TO EST;
--
-- See that the timezone offsets are preserved to minute resolution
SELECT timestamptp_in('2021-01-02T03:04:05-0123');
      timestamptp_in       
---------------------------
 2021-01-02 03:04:05-01:23
(1 row)

--
-- Use cast syntax
SELECT '2010-01-02 03:04:05 +06:07'::timestamptp;
        timestamptp        
---------------------------
 2010-01-02 03:04:05+06:07
(1 row)

SELECT '2000-01-01 00:00:00Z'::timestamptp;
      timestamptp       
------------------------
 2000-01-01 00:00:00+00
(1 row)

--
-- Extract timestamptz and offset from a timestamptp
SELECT '2000-01-01 00:00:00 -07:00'::timestamptp::timestamptz;
         timestamptz          
------------------------------
 Sat Jan 01 02:00:00 2000 EST
(1 row)

SELECT get_offset('2000-01-01 00:00:00 -07:00'::timestamptp);
 get_offset 
------------
       -420
(1 row)

--
-- Construct a timestamptp from a timestamptz and an offset
SELECT make_timestamptp('2000-01-01 00:00:00 +06:00'::timestamptz, -300);
    make_timestamptp    
------------------------
 1999-12-31 13:00:00-05
(1 row)

--
-- Error cases with timezone offset overflow
SELECT '2000-01-01 00:00:00 +16'::timestamptp;
ERROR:  time zone displacement out of range: "2000-01-01 00:00:00 +16"
LINE 1: SELECT '2000-01-01 00:00:00 +16'::timestamptp;
               ^
SELECT '2000-01-01 00:00:00 -16'::timestamptp;
ERROR:  time zone displacement out of range: "2000-01-01 00:00:00 -16"
LINE 1: SELECT '2000-01-01 00:00:00 -16'::timestamptp;
               ^
SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, -960);
ERROR:  Timezone offset must be strictly between -960 and 960
SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, 960);
ERROR:  Timezone offset must be strictly between -960 and 960
--
-- Just under the overflow
SELECT '2000-01-01 00:00:00 +15:59'::timestamptp;
        timestamptp        
---------------------------
 2000-01-01 00:00:00+15:59
(1 row)

SELECT '2000-01-01 00:00:00 -15:59'::timestamptp;
        timestamptp        
---------------------------
 2000-01-01 00:00:00-15:59
(1 row)

SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, -959);
     make_timestamptp      
---------------------------
 1999-12-31 08:01:00-15:59
(1 row)

SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, 959);
     make_timestamptp      
---------------------------
 2000-01-01 15:59:00+15:59
(1 row)

--
-- Cast a timestamptp to timestamptz
SELECT ('2000-01-01 00:00:00 +07:00'::timestamptp)::timestamptz;
         timestamptz          
------------------------------
 Fri Dec 31 12:00:00 1999 EST
(1 row)

--
-- Cast a timestamptz to timestamptp using implicit local timezone
SELECT ('2000-01-01 00:00:00 +00:00'::timestamptz)::timestamptp;
      timestamptp       
------------------------
 1999-12-31 19:00:00-05
(1 row)

--
-- Tests for converting timestamptz to timestamptp in a DST-aware timezone
-- These two timestamps are on different sides of a DST transition, so the result
-- should have different offsets
SELECT make_timestamptp('2022-03-12 12:00:00Z'::timestamptz, 'America/New_York');
    make_timestamptp    
------------------------
 2022-03-12 07:00:00-05
(1 row)

SELECT make_timestamptp('2022-03-13 12:00:00Z'::timestamptz, 'America/New_York');
    make_timestamptp    
------------------------
 2022-03-13 08:00:00-04
(1 row)

--
-- Tests make_timestamptp() using other formats for specifying a timezone
-- as text. Offset spec, timezone abbreviations, and full timezone names should
-- be accepted.
SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, '-05:00'::text);
    make_timestamptp    
------------------------
 1999-12-31 19:00:00-05
(1 row)

SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, '+0500'::text);
    make_timestamptp    
------------------------
 2000-01-01 05:00:00+05
(1 row)

SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, 'EST'::text);
    make_timestamptp    
------------------------
 1999-12-31 19:00:00-05
(1 row)

SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, 'EDT'::text);
    make_timestamptp    
------------------------
 1999-12-31 20:00:00-04
(1 row)

SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, 'PST'::text);
    make_timestamptp    
------------------------
 1999-12-31 16:00:00-08
(1 row)

SELECT make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, 'US/Eastern'::text);
    make_timestamptp    
------------------------
 1999-12-31 19:00:00-05
(1 row)

--
-- Test operators
SELECT '2000-01-01 00:00:00 +00:00'::timestamptp = '2000-01-01 00:00:00 +00:00'::timestamptp;
 ?column? 
----------
 t
(1 row)

SELECT '2000-01-01 00:00:00 +00:00'::timestamptp = '2000-01-02 00:00:00 +00:00'::timestamptp;
 ?column? 
----------
 f
(1 row)

SELECT '2000-01-01 00:00:00 +00:00'::timestamptp = make_timestamptp('2000-01-01 00:00:00 +00:00'::timestamptz, 700);
 ?column? 
----------
 t
(1 row)

--
-- timestamptp = timestamptz
SELECT '2000-01-01 00:00:00 +00:00'::timestamptp = '2000-01-01 00:00:00 +00:00'::timestamptz;
 ?column? 
----------
 t
(1 row)

SELECT '2000-01-01 00:00:00 +05:00'::timestamptp = '2000-01-01 00:00:00 +05:00'::timestamptz;
 ?column? 
----------
 t
(1 row)

SELECT '2000-01-01 00:00:00 +00:00'::timestamptz = '2000-01-01 00:00:00 +00:00'::timestamptp;
 ?column? 
----------
 t
(1 row)

--
-- Test timestamptp + interval
SELECT '2000-01-01 00:00:00 +02:00'::timestamptp + '5 days'::interval;
        ?column?        
------------------------
 2000-01-06 00:00:00+02
(1 row)

SELECT '5 days'::interval + '2000-01-01 00:00:00 +02:00'::timestamptp;
        ?column?        
------------------------
 2000-01-06 00:00:00+02
(1 row)

